### PR TITLE
Clarify LMDB-dumps and MinSaver in Documentation

### PR DESCRIPTION
**Pull request by PatWie**
The user might be familiar with creating a ds-stream and wants to
dump the compressed image to disk as an lmdb file. This adds this to
the documentation but gives a warning about the write-performance.

In addition, there is a small amount of  documentation about the
Min-Saver and ScalarStats-runner, which is slightly now described in
more detail.

### CODE REVIEW

**Comment by PatWie** [comment]
@ppwwyyxx Are you really sure, that the reading speed is just 130 it/s ? I get 400 it/s on local SSD.

File: docs/tutorial/efficient-dataflow.md [hunk 1]
```diff
@@ -157,19 +168,19 @@ Then we add necessary transformations:
 .. code-block:: python
     :emphasize-lines: 3-5
 
-    ds = LMDBData(db, shuffle=False)
+    ds = LMDBData(db, shuffle=False) # produces key-value pairs
     ds = LocallyShuffleData(ds, 50000)
-    ds = LMDBDataPoint(ds)
-    ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
+    ds = LMDBDataPoint(ds)  # converts 'ds' to  pairs (encoded_image, label)
+    ds = ImageDecode(ds, mode='.jpg', index=0)
```
Comment (ppwwyyxx) [inline docs/tutorial/efficient-dataflow.md]: Can we just keep the Map? It's a one-liner anyway, and user can see and choose what decoder they want (I know faster decoder than opencv). The wrappers also lose the options available in imdecode/imencode (IMREAD mode, image quality, etc).

**Comment by ppwwyyxx** [comment]
Probably RAM on my test machines wasn't very good. I just ran a hdparm benchmark on a new machine and see 11G/s cached read speed. This seems to be close to the speed you get.

**Comment by ppwwyyxx** [comment]
No. 130it/s is after batching to 256.

File: docs/tutorial/efficient-dataflow.md (hunk 1 above)
Comment (PatWie) [inline docs/tutorial/efficient-dataflow.md]: `SelectComponent`, `ColorSpace`, `Grayscale`, `ToUint8` , `ToFloat32`, `FixedCrop`, `Clip`, `Identity`, `JpegNoise`, `Resize` are also one-liner.

Comment (ppwwyyxx) [inline docs/tutorial/efficient-dataflow.md]: Yeah I think FixedCrop is especially unnecessary. When I see the function I'll be wondering whether it crops to max+1 or max.

File: tensorpack/dataflow/image.py [hunk 2]
```diff
@@ -11,7 +11,39 @@
...
+class ImageDecode(MapDataComponent):
+    """ Decode image as a string. """
+    def __init__(self, ds, index=0, mode=cv2.IMREAD_COLOR):
+        """
+        Args:
+            ds (DataFlow): input DataFlow.
+            index (int): the index of the component to be decoded.
+        """
+        def func(im_data):
+            return cv2.imdecode(np.asarray(bytearray(im_data), dtype=np.uint8), mode)
```
Comment (ppwwyyxx) [inline tensorpack/dataflow/image.py]: Would bytearray introduce a copy when input is already a uint8 nparray?..

File: docs/tutorial/efficient-dataflow.md (hunk 1 above)
Comment (ppwwyyxx) [inline docs/tutorial/efficient-dataflow.md]: Also, if it is provided as a dataflow called `ImageDecode`, people would want to use it with one datatype or another (at least bytes and nparray are common). Similarly for encode, we cannot assume users want an nparray in the end, bytes can be useful as well.
 Although they seem so, these two functions are not so well-defined like `Clip`,`ToUint8`,`Resize` which you don't need to read docs to know what they do.
I don't mind removing `SelectComponent`, actually I would hope to let users get familiar with `MapData/MapImage` so they won't expect all simple functionalities built into tensorpack.

File: tensorpack/dataflow/image.py (hunk 2 above)
Comment (PatWie) [inline tensorpack/dataflow/image.py]: It does

**Comment by PatWie** [comment]
I would still vote for these atomic on-liners `ImageEncode`, `ImageDecode`  as they are pretty common in any good LMDB based training procedure. Please note, that the code from the current documentation does not work:

```python
ds = LMDBData(db, shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = LMDBDataPoint(ds)
```
In `LMDBDataPoint` there is a `isinstance(args[0], LMDBData)` which is obviously `False` if there is a `LocallyShuffleData` between.

Maybe a renaming from
`LMDBDataPoint` to `MsgPackDecoder` or `LMDBdumpDeoder` is also more intuitiv.

File: tensorpack/dataflow/format.py [hunk 3]
```diff
@@ -160,16 +160,15 @@ class LMDBDataPoint(MapData):
     """ Read a LMDB file and produce deserialized values.
         This can work with :func:`tensorpack.dataflow.dftools.dump_dataflow_to_lmdb`. """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, ds):
         """
         Args:
-            args, kwargs: Same as in :class:`LMDBData`.
-        """
-        if isinstance(args[0], LMDBData):
```
Comment (PatWie) [inline tensorpack/dataflow/format.py]: This is almost never True.

Comment (ppwwyyxx) [inline tensorpack/dataflow/format.py]: Right.

**Comment by ppwwyyxx** [comment]
I agree they are common and I'm only concerned with the type confusion. "Encode" always means "convert to bytes" except in opencv. In fact, encoding to nparray is only useful when you decode with opencv. 
When we add distributed training, we probably have to use jpeg-encoded __bytes__ in a TFRecord and decode with TF, to reach maximum speed. At least that's what google is doing. In that case we'll need encode/decode to handle bytes.
I don't mind adding it if it could at least support bytes, but it will just look too complicated (both the docs and the code) given that it could be done in one line. Originally I used bytes in the tutorial and I only added nparray conversion for speed (and not even sure if it really helps).

I agree on the rename. But `MsgPackDecoder` is not very good because it decodes `dp[1]`, which is an unexpected behavior from the name. Decoding from `dp[1]` is LMDB-specific behavior (just like decoding nparray is cv-specific behavior) so we better keep LMDB in the name.

**Comment by PatWie** [comment]
At least there should be a convenient usage:
- get data
- encode
- dump

and then
- load
- decode

What about a combined `Serializer` class which can return properties like a data stream for reading and writing, like

```
ds = ...
wrapper = MyFileTypeWrapper(file='/tmp/test.lmdb', format='lmdb')
wrapper.serialize(ds)
```

and then

```
wrapper = MyFileTypeWrapper(file='/tmp/test.lmdb', format='lmdb')
wrapper.deserialize(ds) # a combined version of image-encode, decode and loads and dump
```

**Comment by ppwwyyxx** [comment]
I didn't get why this is needed. What's the benefit of using a new API serialize/deserialize compared to the original APIs `dump_to_lmdb` and `LMDBDataPoint`? You need to write decode/encode somewhere anyway.

**Comment by PatWie** [comment]
Closing for now. But you should remove the lines:
https://github.com/ppwwyyxx/tensorpack/blob/master/tensorpack/dataflow/format.py#L168-L171
as we discussed here:
https://github.com/ppwwyyxx/tensorpack/pull/267/files#diff-4a08f729b0314e28cdde668081b6ce22

**Pull request closed without merging**